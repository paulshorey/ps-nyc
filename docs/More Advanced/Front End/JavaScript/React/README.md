# React

### Shorthand (only render method, no lifecycle):

```javascript
const Profile = props => {
    
  // validate inputs
  if (!props.someList || !props.someList.length) {
      return null;
  }
  
  // what are we rendering?
  const Items = [];
  for (let i=0; i<props.someList.length; i++) {
      Items.push(<li>{props.someList[i]}</li>);
  };
  
  // ok, lets render it
  return (<ol>{Items}</ol>);
};
```

## 
### this.state
**Set state only after render(), in componentDidUpdate() and custom methods.**, always being paranoid to avoid infinite loops.  
  
**Also,**  
Because `this.props` and `this.state` may be updated asynchronously, you should not rely on their values for calculating the next state.
```javascript
    /* do this if there's any possibility of vars changing in the meanwhile */
    this.setState((state, props)=>(
        { counter: state.counter + props.increment }
    ));
``` 
https://medium.freecodecamp.org/get-pro-with-react-setstate-in-10-minutes-d38251d1c781

## 
### Version 17 (new and renamed lifecycle methods):
* UNSAFE_componentWillMount
* UNSAFE_componentWillRecieveProps
* UNSAFE_componentWillUpdate
* **getDerivedStateFromProps** (prevProps, prevState) {}
    * This method is going to handle what **componentWillRecieveProps** was able to do along with **componentDidUpdate**. It is static. It is called after a component is created and also called when it receives a new prop. This will be the safer alternative to **componentWillRecieveProps**.
* **getSnapshotBeforeUpdate** (prevProps, prevState) {}
    * This is going to handle what **componentWillUpdate** was able to do along with **componentDidUpdate**. This is called right before the DOM is updated. The value that is returned from **getSnapshotBeforeUpdate** is passed on to **componentDidUpdate**.
