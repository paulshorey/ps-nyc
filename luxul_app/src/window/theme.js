export default (window.theme = {
	topHeight: '4.5rem',
	sidebarNavWidth: '244px',
	fontFamily: 'Agenda, Helvetica, sans-serif',
	containerMaxWidth: '1080px',
	colorNeutral: '#999999',
	colorGreen1: '#81c341',
	colorGreen2: '#6ab81c',
	colorGreen3: '#659520',
	colorBlue1: '#277a80',
	colorBlue2: '#39617a',
	colorBlue3: '#324E68',
	colorOrange1: '#fcca56',
	colorOrange2: '#fdb713',
	colorOrange3: '#f1a700',
	colorWarn1: 'rgb(244, 67, 54)',
	colorWarn2: '#d9534f',
	colorWarn3: '#bf4343',
	colorDark1: 'rgb(0, 0, 0)',
	colorDark2: 'rgb(30,30,30)',
	colorDark3: 'rgb(60,60,60)',
	colorLight1: 'rgb(255, 255, 255)',
	colorLight2: 'rgb(245,245,245)',
	colorLight3: 'rgb(230,230,230)',
	colorOnDark1: 'rgba(255, 255, 255, 0.9)',
	colorOnDark2: 'rgba(255, 255, 255, 0.6)',
	colorOnDark3: 'rgba(255, 255, 255, 0.3)',
	colorOnLight1: 'rgba(0, 0, 0, 0.8)',
	colorOnLight2: 'rgba(0, 0, 0, 0.5)',
	colorOnLight3: 'rgba(0, 0, 0, 0.2)',
});
