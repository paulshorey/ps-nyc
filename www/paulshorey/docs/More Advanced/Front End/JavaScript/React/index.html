<!doctype html>
<html>

<head>

    <meta charset="utf-8">
    <meta name="viewport" content="initial-scale=1, maximum-scale=1, user-scalable=no, width=device-width">
    <link href="/assets/gfx/puzzle.ico" rel="shortcut icon" type="image/x-icon"/>

    <link rel="stylesheet" type="text/css" href="/docs/template.css"/>
</head>

<body>
<div class="page">
    <nav class="nav">
        <ul>
            <li class="heading">
                <a class="title-items" href="/docs"><span class="emoji">üìù</span><span class="emoji-text">/docs</span></a>
            </li>
            <ul class=""><li class="heading isActive"><span>Beginner</span></li>
<ul class=""><li class="heading isActive"><span>Front End</span></li>
<ul class=""><li class="heading "><a href="../../../../Beginner/Front%20End/CSS/index.html">CSS</a></li>

<li class="heading "><a href="../../../../Beginner/Front%20End/JavaScript/index.html">JavaScript</a></li>

</ul>
</ul>
<li class="heading isActive"><span>More Advanced</span></li>
<ul class=""><li class="heading isActive"><span>Basics</span></li>
<ul class=""><li class="heading "><a href="../../../Basics/CLI/index.html">CLI</a></li>
<ul class=""><li class=""><a href="../../../Basics/CLI/Nice-Editor.html">Nice Editor</a></li><li class=""><a href="../../../Basics/CLI/shortcuts.html">shortcuts</a></li></ul>
<li class="heading "><a href="../../../Basics/GIT%20+%20SSH/index.html">GIT + SSH</a></li>
<ul class=""><li class=""><a href="../../../Basics/GIT%20+%20SSH/porcelain.html">porcelain</a></li></ul>
<li class="heading "><a href="../../../Basics/IDE%20+%20Code/index.html">IDE + Code</a></li>
<ul class=""><li class=""><a href="../../../Basics/IDE%20+%20Code/auto-formatting.html">auto formatting</a></li><li class=""><a href="../../../Basics/IDE%20+%20Code/edit-over-SSH.html">edit over SSH</a></li></ul>
<li class="heading "><a href="../../../Basics/RegEx/index.html">RegEx</a></li>
<ul class=""><li class=""><a href="../../../Basics/RegEx/Basics.html">Basics</a></li><li class=""><a href="../../../Basics/RegEx/HTML.html">HTML</a></li><li class=""><a href="../../../Basics/RegEx/Javascript.html">Javascript</a></li><li class=""><a href="../../../Basics/RegEx/Markdown.html">Markdown</a></li></ul>
</ul>
<li class="heading isActive"><span>Front End</span></li>
<ul class=""><li class="heading "><a href="../../CSS/index.html">CSS</a></li>
<ul class=""><li class=""><a href="../../CSS/frameworks.html">frameworks</a></li></ul>
<li class="heading childIsActive"><a href="../index.html">JavaScript</a></li>
<ul class="isActive"><li class=""><a href="../ES6.html">ES6</a></li><li class="heading isActive"><a href="./index.html">React</a></li>

<li class=""><a href="../looping.html">looping</a></li><li class=""><a href="../sorting.html">sorting</a></li></ul>
</ul>
</ul>
<li class="heading isActive"><span>SysAdmin</span></li>
<ul class=""><li class="heading isActive"><span>Containers</span></li>
<ul class=""><li class="heading "><a href="../../../../SysAdmin/Containers/Docker/index.html">Docker</a></li>
<ul class=""><li class=""><a href="../../../../SysAdmin/Containers/Docker/docker-compose.html">docker compose</a></li><li class=""><a href="../../../../SysAdmin/Containers/Docker/docker.html">docker</a></li><li class=""><a href="../../../../SysAdmin/Containers/Docker/help%20docker-compose.html">help docker compose</a></li><li class=""><a href="../../../../SysAdmin/Containers/Docker/help%20docker-run.html">help docker run</a></li><li class=""><a href="../../../../SysAdmin/Containers/Docker/help%20docker.html">help docker</a></li></ul>
</ul>
<li class="heading isActive"><span>Linux</span></li>
<ul class=""><li class=""><a href="../../../../SysAdmin/Linux/easy-deployment.html">easy deployment</a></li><li class=""><a href="../../../../SysAdmin/Linux/linux-from-scratch.html">linux from scratch</a></li></ul>
<li class=""><a href="../../../../SysAdmin/Ports.html">Ports</a></li><li class=""><a href="../../../../SysAdmin/SSL.html">SSL</a></li></ul>
</ul>
        </ul>
    </nav>
    <main class="content">
        <h1>React</h1>
<h3>Shorthand (only render method, no lifecycle):</h3>
<pre><code class="language-javascript prettyprint">const Profile = props =&gt; {
    
  // validate inputs
  if (!props.someList || !props.someList.length) {
      return null;
  }
  
  // what are we rendering?
  const Items = [];
  for (let i=0; i&lt;props.someList.length; i++) {
      Items.push(&lt;li&gt;{props.someList[i]}&lt;/li&gt;);
  };
  
  // ok, lets render it
  return (&lt;ol&gt;{Items}&lt;/ol&gt;);
};
</code></pre>
<h2></h2>
<h3>this.state</h3>
<p><strong>Set state only after render(), in componentDidUpdate() and custom methods.</strong>, always being paranoid to avoid infinite loops.</p>
<p><strong>Also,</strong><br>
Because <code>this.props</code> and <code>this.state</code> may be updated asynchronously, you should not rely on their values for calculating the next state.</p>
<pre><code class="language-javascript prettyprint">    /* do this if there's any possibility of vars changing in the meanwhile */
    this.setState((state, props)=&gt;(
        { counter: state.counter + props.increment }
    ));
</code></pre>
<p><a href="https://medium.freecodecamp.org/get-pro-with-react-setstate-in-10-minutes-d38251d1c781">https://medium.freecodecamp.org/get-pro-with-react-setstate-in-10-minutes-d38251d1c781</a></p>
<h2></h2>
<h3>Version 17 (new and renamed lifecycle methods):</h3>
<ul>
<li>UNSAFE_componentWillMount</li>
<li>UNSAFE_componentWillRecieveProps</li>
<li>UNSAFE_componentWillUpdate</li>
<li><strong>getDerivedStateFromProps</strong> (prevProps, prevState) {}
<ul>
<li>This method is going to handle what <strong>componentWillRecieveProps</strong> was able to do along with <strong>componentDidUpdate</strong>. It is static. It is called after a component is created and also called when it receives a new prop. This will be the safer alternative to <strong>componentWillRecieveProps</strong>.</li>
</ul>
</li>
<li><strong>getSnapshotBeforeUpdate</strong> (prevProps, prevState) {}
<ul>
<li>This is going to handle what <strong>componentWillUpdate</strong> was able to do along with <strong>componentDidUpdate</strong>. This is called right before the DOM is updated. The value that is returned from <strong>getSnapshotBeforeUpdate</strong> is passed on to <strong>componentDidUpdate</strong>.</li>
</ul>
</li>
</ul>

    </main>
</div>
</body>
<script src="https://cdn.jsdelivr.net/gh/google/code-prettify@master/loader/run_prettify.js"></script>

</html>