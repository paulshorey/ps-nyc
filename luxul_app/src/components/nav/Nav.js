// @flow
import React, { Component } from 'react';
import { Link, NavLink } from 'react-router-dom';
import NavGroup from './NavGroup';
import * as Styled from './NavStyled';

type PropsHOC = {
	routes: [],
	currentPath: {},
	children: {},
};
type StateHOC = {
	nav: {},
};
type Props = {
	routes: [], // list of routes/urls/components from src/devices
	currentPath: {}, // window.location.pathname ... but more React'ive
	nav: {}, // window.store.nav ... passed from NavConnected component below
};

export class Nav extends Component<Props> {
	renderLinks = (links: any) => {
		return links.map((link, index) => {
			return (
				<NavLink
					key={link.url + index}
					className="title link"
					activeClassName="active"
					to={link.url}
					onClick={()=>{
						document.getElementById('Hamburger').classList.remove('opened');
						window.store.nav={}
					}}
				>
					<span className="fontIcon icon-navlink_dot" />
					<span>{link.title}</span>
				</NavLink>
			);
		});
	};

	renderNav() {
		const { routes, currentPath } = this.props;
		return routes.map((link, index) => {
			if (link.children) {
				return (
					<NavGroup key={link.title + index} title={link.title} currentPath={currentPath}>
						{this.renderLinks(link.children)}
					</NavGroup>
				);
			} else {
				return this.renderLinks([link]);
			}
		});
	}

	render() {
		return (
			<Styled.Nav className={'nav_left' + (this.props.nav.opened ? '' : ' closed')}>
				{this.renderNav()}

				<Link className="title link" to="/">
					<span className="fontIcon icon-navlink_dot" />
					<span>Logout</span>
				</Link>
			</Styled.Nav>
		);
	}
}

/*
	when {window.store.nav} changes, update {this.props.nav} in <Nav /> component above
*/
class NavConnected extends React.Component<PropsHOC, StateHOC> {
	state = {
		nav: window.store.nav,
	};
	componentWillMount() {
		window.store.watch('nav', (name, oldValue, value) => {
			this.setState({ [name]: value });
			return value;
		});
	}
	componentWillUnmount() {
		window.store.unwatch('nav');
	}
	render() {
		return <Nav {...this.props} nav={this.state.nav} />;
	}
}

export default NavConnected;
