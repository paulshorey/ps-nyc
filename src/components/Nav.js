import React from "react"
import { FontAwesomeIcon as FA } from "@fortawesome/react-fontawesome"
import { faSmileBeam, faPhotoVideo } from "@fortawesome/pro-light-svg-icons"
import { faPhone, faFilePdf, faEnvelope } from "@fortawesome/pro-solid-svg-icons"
// import styled from "styled-components"

class Header extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      phone: "",
      email: "",
      resume: ""
    }
  }
  componentDidMount() {
    if (typeof window === "object" && "$" in window && !this.state.phone) {
      this.setState({
        phone: "385.770.6789",
        email: "hello@paulshorey.com",
        resume: "/file.docx"
      })
    }
    if (typeof window === "object" && "$" in window) {
      let $ = window.$
      $("[secretmailto]").each(function () {
        $(this).get(0).innerHTML = $(this).attr("secretmailto")
        $(this).attr("href", "mailto:" + $(this).attr("secretmailto"))
      })

      $("[secrettel]").each(function () {
        $(this).get(0).innerHTML = $(this).attr("secrettel")
        $(this).attr("href", "tel:" + $(this).attr("secrettel").replace(/./g, ""))
      })
    }
  }
  componentDidUpdate() {
    if (typeof window === "object" && "$" in window && !this.state.phone) {
      this.setState({
        phone: "385.770.6789",
        email: "hello@paulshorey.com",
        resume: "/file.docx"
      })
    }
  }
  render() {
    return (
      <header className="content">
        <div className="h-left">
          <span>
            <a href="/about">
              <span className="icon">
                &thinsp;
                <FA icon={faSmileBeam} style={{ marginTop: "0.25rem" }} />
              </span>
              Paul Shorey
            </a>{" "}
            <span>&nbsp;|&nbsp;Javascript Software Engineer / UI Designer / Tinkerer &nbsp;</span>
          </span>
          <a href="/about">
            <span className="icon">
              <FA icon={faPhotoVideo} className="x85" style={{ marginTop: "0.25rem" }} />
            </span>
            Photos | Videos | Hobbies
          </a>{" "}
          {/*<a href="/about">*/}
          {/*  <span className="icon">*/}
          {/*    <FA icon={faPhotoVideo} className="x85" style={{ marginTop: "0.25rem" }} />*/}
          {/*  </span>*/}
          {/*  Projects*/}
          {/*</a>{" "}*/}
        </div>
        <div className="h-right">
          <a href={"mailto:" + this.state.email}>
            {this.state.email} <FA icon={faEnvelope} className="x85" style={{ marginTop: "0.25rem" }} />
          </a>{" "}
          &nbsp;|&nbsp;
          <a href={"tel:" + this.state.phone}>
            {this.state.phone} <FA icon={faPhone} className="x85" />
          </a>
          &nbsp;|&nbsp;
          <a href={this.state.resume}>
            Resume &nbsp;
            <FA icon={faFilePdf} className="x85" />
          </a>
        </div>
      </header>
    )
  }
}

export default Header
