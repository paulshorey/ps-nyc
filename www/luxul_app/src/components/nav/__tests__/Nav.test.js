import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router, withRouter } from 'react-router-dom';
import theme from 'window/theme.js';
import Nav from '../Nav';

const routes = [
	{
		title: 'Status',
		url: '/XWC1000/status',
		children: [
			{ title: 'Controller', url: '/XWC1000/status/controller' },
			{ title: 'AP', url: '/XWC1000/status/ap' },
		],
	},
	{
		title: 'Profile Management',
		url: '/XWC1000/profiles',
		children: [
			{ title: 'AP Management', url: '/XWC1000/profiles/ap_management' },
			{ title: 'AP Assignment', url: '/XWC1000/profiles/ap_assignment' },
			{ title: 'SSID Management', url: '/XWC1000/profiles/ssid_management' },
			{ title: 'Wireless Settings', url: '/XWC1000/profiles/wireless_settings' },
		],
	},
	{ title: 'VLAN', url: '/XWC1000/vlan' },
	{ title: 'Controller Settings', url: '/XWC1000/controller' },
];

const history = [
	{
		title: 'Status',
		url: '/XWC1000/status',
		children: [
			{ title: 'Controller', url: '/XWC1000/status/controller' },
			{ title: 'AP', url: '/XWC1000/status/ap' },
		],
	},
	{
		title: 'Profile Management',
		url: '/XWC1000/profiles',
		children: [
			{ title: 'AP Management', url: '/XWC1000/profiles/ap_management' },
			{ title: 'AP Assignment', url: '/XWC1000/profiles/ap_assignment' },
			{ title: 'SSID Management', url: '/XWC1000/profiles/ssid_management' },
			{ title: 'Wireless Settings', url: '/XWC1000/profiles/wireless_settings' },
		],
	},
	{ title: 'VLAN', url: '/XWC1000/vlan' },
	{ title: 'Controller Settings', url: '/XWC1000/controller' },
];

it('renders without crashing', () => {
	const div = document.createElement('div');

	ReactDOM.render(
		<Router>
			<Nav
				history={history}
				routes={routes}
				nav={window.store.opened}
				currentPath="/XWC1000/profiles/ap_management"
			/>
		</Router>,
		div
	);
});

test('theme snapshot is correct', () => {
	expect(theme).toMatchSnapshot();
});
