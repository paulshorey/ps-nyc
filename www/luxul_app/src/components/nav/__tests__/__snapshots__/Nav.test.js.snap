// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theme snapshot is correct 1`] = `
Object {
  "colorBlue1": "#277a80",
  "colorBlue2": "#39617a",
  "colorBlue3": "#324E68",
  "colorDark1": "rgb(0, 0, 0)",
  "colorDark2": "rgb(30,30,30)",
  "colorDark3": "rgb(60,60,60)",
  "colorGreen1": "#81c341",
  "colorGreen2": "#6ab81c",
  "colorGreen3": "#659520",
  "colorLight1": "rgb(255, 255, 255)",
  "colorLight2": "rgb(245,245,245)",
  "colorLight3": "rgb(230,230,230)",
  "colorNeutral": "#999999",
  "colorOnDark1": "rgba(255, 255, 255, 0.9)",
  "colorOnDark2": "rgba(255, 255, 255, 0.6)",
  "colorOnDark3": "rgba(255, 255, 255, 0.3)",
  "colorOnLight1": "rgba(0, 0, 0, 0.8)",
  "colorOnLight2": "rgba(0, 0, 0, 0.5)",
  "colorOnLight3": "rgba(0, 0, 0, 0.2)",
  "colorOrange1": "#fcca56",
  "colorOrange2": "#fdb713",
  "colorOrange3": "#f1a700",
  "colorWarn1": "rgb(244, 67, 54)",
  "colorWarn2": "#d9534f",
  "colorWarn3": "#bf4343",
  "containerMaxWidth": "1080px",
  "fontFamily": "Agenda, Helvetica, sans-serif",
  "sidebarNavWidth": "244px",
  "topHeight": "4.5rem",
}
`;
